core/market/decentralized/src/matcher.rs

```
@@ -140,60 +181,334 @@ impl Matcher {
-            SaveOfferError::Expired(_) => Ok(Propagate::No(Reason::Expired)),
-            // Below errors are not possible to get from checked_store_offer
-            SaveOfferError::SaveError(_, _)
-            | SaveOfferError::SubscriptionValidation(_)
-            | SaveOfferError::WrongState { .. } => {
-                Ok(Propagate::No(Reason::Error(format!("{}", e))))
+        .map_err(|e| log::warn!("Error filtering Offers. Error: {}", e))?)
+}
+
+pub(crate) async fn on_offers_received(
+    resolver: Resolver,
+    caller: String,
+    msg: OffersReceived,
+) -> Result<Vec<SubscriptionId>, ()> {
+    let added_offers_ids = futures::stream::iter(msg.offers.into_iter())
+        .filter_map(|offer| {
+            let resolver = resolver.clone();
+            let offer_id = offer.id.clone();
+            async move {
+                resolver
+                    .store
+                    .save_offer(offer)
+                    .await
+                    .map(|offer| {
+                        resolver.receive(&offer);
+                        offer.id
+                    })
+                    .map_err(|e| {
+                        log::warn!("Failed to save Offer [{}]. Error: {}", &offer_id, &e);
+                        e
+                    })
+                    .ok()
             }
         })
+        .collect::<Vec<SubscriptionId>>()
+        .await;
+
+    log::info!(
+        "Received new Offers from [{}]: \n{}",
+        caller,
+        DisplayVec(&added_offers_ids)
+    );
+    Ok(added_offers_ids)
+}
+
+pub(crate) async fn on_get_offers(
+    resolver: Resolver,
+    _caller: String,
+    msg: GetOffers,
+) -> Result<Vec<Offer>, DiscoveryRemoteError> {
+    match resolver.store.get_offers_batch(msg.offers).await {
+        Ok(offers) => Ok(offers),
+        Err(e) => {
+            log::error!("Failed to get batch offers. Error: {}", e);
+            // TODO: Propagate error.
+            Ok(vec![])
+        }
```

Return error instead of empty vector